### PR TITLE
S 62/el 2610

### PR DESCRIPTION
**Pull request by AlastairMcKee**
Added pdf export and documentation/plunker

### CODE REVIEW

File: docs/app/pages/components/sections/pdf-export/pdf-export/pdf-export.component.html [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+                            chartType="bar">
+                    </canvas>
+
+                </div>
+            </div>
+        </div>
+    </ux-dashboard-widget>
+
+    <ng-template pdfExportItem row="2" columns="3" let-data="data" [data]="dataTable">
+        <table class="m-md" style="width: 200px;">
```
Comment (ashley-hunter) [inline docs/app/pages/components/sections/pdf-export/pdf-export/pdf-export.component.html]: Move all the inline styles to classes

File: docs/app/pages/components/sections/pdf-export/pdf-export/pdf-export.component.html [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+        </div>
+    </ux-dashboard-widget>
+
+</ux-dashboard>
+
+<button class="btn btn-secondary" (click)="getDocument()">Get Document</button>
+
+<hr>
+
+<p>The <code>$PdfExportService</code> can be used to export sections of a webpage to be saved as PDF or printed.</p>
```
Comment (ashley-hunter) [inline docs/app/pages/components/sections/pdf-export/pdf-export/pdf-export.component.html]: PdfExportService shouldnt have a $ before it

File: docs/app/pages/components/sections/pdf-export/pdf-export/snippets/app.ts [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+
+    options: DashboardOptions = {
+        columns: 3,
+        padding: 10,
+        rowHeight: 220,
+        emptyRow: false,
+        minWidth: 187
+    };
+
+    constructor(public colorService: ColorService, private _pdfExportService: PdfExportService) {
```
Comment (ashley-hunter) [inline docs/app/pages/components/sections/pdf-export/pdf-export/snippets/app.ts]: colorService is only used within the constructor so it doesn't need to be added as an instance variable, (make same changes in the non-snippet version of this class too)

File: src/components/pdf-export/pdf-export.service.ts [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        });
+
+        htmlElement.appendChild(head);
+        htmlElement.appendChild(body);
+
+        return htmlElement;
+    }
+
+    private cloneHead(): HTMLHeadElement {
+        let head = document.head.cloneNode(true) as HTMLHeadElement;
```
Comment (ashley-hunter) [inline src/components/pdf-export/pdf-export.service.ts]: Simplify just to:

```typescript
return document.head.cloneNode(true) as HTMLHeadElement;
```

File: src/components/pdf-export/pdf-export.service.ts [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        return head;
+    }
+
+    private clone(element: HTMLElement): HTMLElement {
+        return element instanceof HTMLCanvasElement ? this.cloneCanvas(element)
```
Comment (ashley-hunter) [inline src/components/pdf-export/pdf-export.service.ts]: Change this to a if statement. Try to avoid nested ternary if statements as they get hard to follow very quickly.

File: src/components/pdf-export/pdf-export.service.ts [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        return head;
+    }
+
+    private clone(element: HTMLElement): HTMLElement {
+        return element instanceof HTMLCanvasElement ? this.cloneCanvas(element) 
+        : element.tagName === 'ng-template' ? this.template(element)
+        : this.cloneElement(element);
+    }
+
+    private template(element: HTMLElement): HTMLElement {
```
Comment (ashley-hunter) [inline src/components/pdf-export/pdf-export.service.ts]: Is this function needed?

**Comment by ashley-hunter** [comment]
Added Donna as a reviewer too - I have had a look over the code, put I'm not fully aware of all the requirements of this feature so would be good to get a second set of eyes looking over it.

**Comment by AlastairMcKee** [comment]
Made requested changes

File: src/components/pdf-export/pdf-export.service.ts [hunk 7]
```diff
@@ -70,19 +70,18 @@ export class PdfExportService {
...
-        : element.tagName === 'ng-template' ? this.template(element)
-        : this.cloneElement(element);
-    }
 
-    private template(element: HTMLElement): HTMLElement {
-        let x = element;
-        return element;
+        if (element instanceof HTMLCanvasElement) {
+            return this.cloneCanvas(element);
+        } else if (element.tagName === 'ng-template') {
```
Comment (ashley-hunter) [inline src/components/pdf-export/pdf-export.service.ts]: Tag name returns a string with uppercase letters, will need to account for that. You should also possible account here for a `template` tag which is essentially the same as `ng-template` tags

**Comment by AlastairMcKee** [comment]
Made the change requested by Ashley, the check for template is no longer required so i have simplified things a little. Some of the points mentioned by Donna are deliberate, possibly poorly documented how ever there are still some issues I have missed, I will try to look into these before the meeting.

Thanks

**Comment by donna-marijne** [comment]
As discussed - please set up an internal repo with this work and a summary of any relevant technical details, to be continued when new requirements are available. Thanks.
